src/main/java/com/mixpanel/android/viewcrawler/ViewVisitor.java

```diff
@@ -152,19 +180,111 @@ public void accumulate(View found) {
+                throw e;
+            }
         }
 
         private void setLayout(View target, int verb, int anchorId) {
             RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)target.getLayoutParams();
             params.addRule(verb, anchorId);
+
+            final ArrayList<Integer> rules;
+            if (mHorizontalRules.contains(verb)) {
+                rules = mHorizontalRules;
+            } else if (mVerticalRules.contains(verb)) {
+                rules = mVerticalRules;
+            } else {
+                rules = null;
+            }
+
+            if (rules != null && !verifyLayout(target, rules)) {
+                JSONObject errorInfo = new JSONObject();
+                try {
+                    errorInfo.put("cid", mName);
+                    errorInfo.put("error_type", "circular_dependency");
+                } catch (JSONException e) {
+                    ; // won't reach here
+                }
+                throw new LayoutUpdateException("Circular dependency detected!", errorInfo);
+            }
+
             target.setLayoutParams(params);
         }
 
+        private boolean verifyLayout(View target, ArrayList<Integer> rules) {
+            ViewGroup parent = (ViewGroup) target.getParent();
+            SparseArray<View> idToChild = new SparseArray<View>();
+
+            for (int i = 0; i < parent.getChildCount(); i++) {
+                View child = parent.getChildAt(i);
+                idToChild.put(child.getId(), child);
+            }
+
+            ArrayMap<View, ArrayList<View>> dependencyGraph = new ArrayMap<View, ArrayList<View>>();
+            for (int i = 0; i < idToChild.size(); i++) {
+                final View child = idToChild.valueAt(i);
+                final RelativeLayout.LayoutParams childLayoutParams = (RelativeLayout.LayoutParams) child.getLayoutParams();
+                int[] layoutRules = childLayoutParams.getRules();
+
+                ArrayList<View> dependencies = new ArrayList<View>();
+                for (int rule : rules) {
+                    int dependencyId = layoutRules[rule];
+                    if (dependencyId > 0 && dependencyId != child.getId()) {
+                        dependencies.add(idToChild.get(dependencyId));
+                    }
+                }
+
+                dependencyGraph.put(child, dependencies);
+            }
+
+            return hasCycle(dependencyGraph);
+        }
+
+        private boolean hasCycle(ArrayMap<View, ArrayList<View>> dependencyGraph) {
+            ArrayList<View> dfsStack = new ArrayList<View>();
+            while (!dependencyGraph.isEmpty()) {
+                View currentNode = dependencyGraph.keyAt(0);
+                if (!subGraphHasCycle(dependencyGraph, currentNode, dfsStack)) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private boolean subGraphHasCycle(ArrayMap<View, ArrayList<View>> dependencyGraph,
```

Comment:
What is the advantage here? I think ArrayMap has all the functionality we need and this code is pretty straight forward. I don't know the advantage of having a class for this thing over doing it inline.